File: dsajava/src/main/java/com/leetcode/dsajava/my_daily_leetcode_challenge/backtrack_recursive/WordLadder_127.java
package com.leetcode.dsajava.my_daily_leetcode_challenge.backtrack_recursive;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class WordLadder_127 {

    public static void main(String[] args) {

    }

    // Time: O(m*n^2), Space: O(m*n^2), n: wordList.size, m: word.length
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        // B1: convert words to adj list
        if (wordList.isEmpty() || !wordList.contains(endWord)) return 0;
        Map<String, Set<String>> adj = new HashMap<>();
        for (int i = 0; i < wordList.size(); i++) {
            for (int j = i + 1; j < wordList.size(); j++) {
                String word1 = wordList.get(i);
                String word2 = wordList.get(j);
                adj.computeIfAbsent(word1, k -> new HashSet<>());
                adj.computeIfAbsent(word2, k -> new HashSet<>());
                if (different(word1, word2) == 1) { // can transform
                    adj.get(word1).add(word2);
                    adj.get(word2).add(word1);
                }
            }
        }
        Set<String> visited = new HashSet<>();
        ArrayList<Pair<String, Integer>> queue = new ArrayList<>();
        // Step 2: beginWord is not in wordList, have to find a word that beginWord can transform to
        for (String word : wordList) {
            if (different(beginWord, word) == 1) { // can transform
                queue.add(new Pair<>(word, 1)); // word, level
                visited.add(word);
            }
        }

        // Step3:
        int i = 0;
        while (i < queue.size()) {
            String word = queue.get(i).fst; // first
            Integer level = queue.get(i).snd; // second
            i++;
            if (word.equals(endWord)) return level + 1;
            else {
                for (String transformation : adj.get(word)) {
                    if(!visited.contains(transformation)) {
                        queue.add(new Pair<>(transformation, level+1));
                        visited.add(transformation);
                    }
                }
            }
        }
        return 0;
    }


    private int different(String word1, String word2) {
        int counter = 0;
        for (int i = 0; i < word1.length(); i++) {
            if (word1.charAt(i) != word2.charAt(i)) counter++;
        }
        return counter;
    }

    public class Pair<A, B> {
        public final A fst;
        public final B snd;

        public Pair(A fst, B snd) {
            this.fst = fst;
            this.snd = snd;
        }
    }
}

File: dsa-kotlin/src/main/java/com/leetcode/dsakotin/my_daily_leetcode/backtrack_recursive/WordLadder_127.kt
package com.leetcode.dsakotin.my_daily_leetcode.backtrack_recursive

fun main() {
    val beginWord = "hit"
    val endWord = "cog"
    val wordList = arrayListOf("hot", "dot", "dog", "lot", "log", "cog")
    println("res: ${ladderLength(beginWord, endWord, wordList)}")
}

// Time: O(m*n^2), Space: O(m*n^2), n: wordList.size, m: word.length
fun ladderLength(beginWord: String, endWord: String, wordList: List<String>): Int {
    if (wordList.isEmpty() || !wordList.contains(endWord)) return 0
    val adj: MutableMap<String, MutableSet<String>> = mutableMapOf()
    // Step 1: convert words to adj list
    for (i in wordList.indices) {
        for (j in i + 1 until wordList.size) {
            val word1 = wordList[i]
            val word2 = wordList[j]
            if (adj[word1] == null) adj[word1] = mutableSetOf()
            if (adj[word2] == null) adj[word2] = mutableSetOf()
            if (difference(word1, word2) == 1) { // can transform
                adj[word1]?.add(word2)
                adj[word2]?.add(word1)
            }
        }
    }

    // Step 2: beginWord is not in wordList, have to find a word that beginWord can transform to
    val queue = arrayListOf<Pair<String, Int>>()
    val visited = mutableSetOf<String>()
    for (word in wordList) {
        if (difference(beginWord, word) == 1) {
            queue.add(Pair(word, 1))
            visited.add(word)
        }
    }

    // Step 3: traverse word list, do transformation
    var i = 0
    while (i < queue.size) {
        val word = queue[i].first
        val level = queue[i].second
        i++
        if (word == endWord) return level + 1
        for (transformation in adj[word].orEmpty()) {
            if (!visited.contains(transformation)) {
                queue.add(Pair(transformation, level + 1))
                visited.add(transformation)
            }
        }
    }
    return 0
}

fun difference(word1: String, word2: String): Int {
    var counter = 0
    word1.indices.forEach { i ->
        if (word1[i] != word2[i]) counter++
    }
    return counter
}